internal/injector/parse/parser.go

```
@@ -0,0 +1,204 @@
+	fset *token.FileSet // thread-safe data structure
+
+	// rawFiles is an intermediary data structure to store the raw content of the files before parsing them.
+	rawFiles []rawFile
+
+	// filesBytesCount is the sum of the bytes of all files parsed so far.
+	filesBytesCount atomic.Uint64
+
+	// eagerness is a flag that is set to true if at least one file has been parsed and requires all files to be parsed.
+	eagerness atomic.Bool
+
+	// parsedFiles is what is returned by ParseFiles.
+	parsedFiles []File
+
+	wg sync.WaitGroup
+
+	errs   []error
+	errsMu sync.Mutex
+}
+
+func NewParser(fset *token.FileSet, nbFiles int) *Parser {
+	p := &Parser{
+		fset:        fset,
+		rawFiles:    make([]rawFile, nbFiles),
+		parsedFiles: make([]File, nbFiles),
+	}
+
+	p.eagerness.Store(true)
+
+	return p
+}
+
+// ParseFiles return either zero files or all files parsed with their respective aspects
+func (p *Parser) ParseFiles(files []string, aspects []*aspect.Aspect) ([]File, error) {
+	p.wg.Add(len(files))
+
+	for idx, file := range files {
+		fileAspects := make([]*aspect.Aspect, len(aspects))
+		copy(fileAspects, aspects)
+
+		go func(idx int, file string) {
+			defer p.wg.Done()
+			var err error
+			p.rawFiles[idx], err = readFile(file)
+			if err != nil {
+				p.addError(fmt.Errorf("reading %q: %w", file, err))
+				return
+			}
+
+			p.filesBytesCount.Add(uint64(len(p.rawFiles[idx].content)))
+			if p.aspectMayMatch() {
+				fileAspects = p.fileFilterAspects(fileAspects, p.rawFiles[idx])
+				if len(fileAspects) == 0 {
+					// No aspects can match on this file, no need to fill up the File.AstFile field.
+					p.parsedFiles[idx] = File{Name: file}
+					return
+				}
+			}
+
+			p.eagerness.Store(false)
+			p.parsedFiles[idx] = p.parseFile(p.rawFiles[idx], fileAspects)
+		}(idx, file)
+	}
+
+	p.wg.Wait()
+
+	// No aspects can match on this package, return nothing
+	if !p.aspectMayMatch() {
+		return nil, nil
+	}
+
+	if len(p.errs) > 0 {
+		return nil, errors.Join(p.errs...)
```

Here, if we have matched 0 aspects; but have encountered at least 1 error (perhaps preventing us from determining we'd have needed to apply some aspect), we silently ignore all the errors?

---

WDYM ?